src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/SaveTempDataPropertyFilterBase.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public abstract class SaveTempDataPropertyFilterBase : ISaveTempDataCallback
+    {
+        protected const string Prefix = "TempDataProperty-";
+
+        protected readonly ITempDataDictionaryFactory _factory;
+
+        /// <summary>
+        /// Describes the temp data properties which exist on <see cref="Subject"/>
+        /// </summary>
+        public IList<TempDataProperty> TempDataProperties { get; set; }
+
+        /// <summary>
+        /// The <see cref="object"/> which has the temp data properties.
+        /// </summary>
+        public object Subject { get; set; }
+
+        /// <summary>
+        /// Tracks the values which originally existed in temp data.
+        /// </summary>
+        public IDictionary<PropertyInfo, object> OriginalValues { get; set; }
+
+        public SaveTempDataPropertyFilterBase(ITempDataDictionaryFactory factory)
+        {
+            _factory = factory;
+        }
+
+        /// <summary>
+        /// Puts the modified values of <see cref="Subject"/> into <paramref name="tempData"/>.
+        /// </summary>
+        /// <param name="tempData">The <see cref="ITempDataDictionary"/> to be updated.</param>
+        public void OnTempDataSaving(ITempDataDictionary tempData)
+        {
+            if (Subject != null && OriginalValues != null)
+            {
+                foreach (var kvp in OriginalValues)
```

Comment:
Do we need to keep track of the OriginalValues at all? Couldn't we just iterate over our TempDataProperties and apply them to tempData without knowing or caring if it changed? Is the cost of setting a tempData value higher than I think?